graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,372 @@
+    // root non-object fields can get explicit hints from their
+    // definition on the Query/Mutation object, if that doesn't exist
+    // then there's no parent field that would assign the default
+    // maxAge, so we do it here.)
+    //
+    // You can disable this on a non-root field by writing
+    // `@cacheControl(inheritMaxAge: true)` on it. If you do this,
+    // then its children will be treated like root paths, since there
+    // is no parent maxAge to inherit.
+
+    if (!fieldPolicy.hasMaxAge()
+        && ((unwrappedReturnType instanceof GraphQLCompositeType && !inheritMaxAge)
+            || parameters.getExecutionStepInfo().getPath().isRootPath())) {
+      fieldPolicy.restrict(defaultMaxAge);
+    }
+
+    return new SimpleInstrumentationContext<ExecutionResult>() {
+      @Override
+      public void onCompleted(ExecutionResult result, Throwable t) {
+        state.restrictCachePolicy(fieldPolicy);
+        super.onCompleted(result, t);
+      }
+    };
+  }
+
+  @Override
+  public CompletableFuture<ExecutionResult> instrumentExecutionResult(
+      ExecutionResult executionResult, InstrumentationExecutionParameters parameters) {
+    final CacheControlState state = parameters.getInstrumentationState();
+
+    if (!state.getOverallPolicyString().isPresent()) {
+      return super.instrumentExecutionResult(executionResult, parameters);
+    }
+
+    if (!executionResult.getErrors().isEmpty()) {
+      return super.instrumentExecutionResult(executionResult, parameters);
+    }
+
+    ExecutionResultImpl resultWithCacheControl =
+        ExecutionResultImpl.newExecutionResult()
+            .from(executionResult)
+            .addExtension(EXTENSION_KEY, state.getOverallPolicyString().get())
+            .build();
+
+    return CompletableFuture.completedFuture(resultWithCacheControl);
+  }
+
+  enum CacheControlScope {
+    PUBLIC,
+    PRIVATE
+  }
+
+  private static class CacheControlState implements InstrumentationState {
+    private final CacheControlPolicy overallPolicy = new CacheControlPolicy();
+
+    public Optional<String> getOverallPolicyString() {
+      return overallPolicy.getPolicy();
+    }
+
+    public void restrictCachePolicy(CacheControlPolicy policy) {
+      this.overallPolicy.restrict(policy);
+    }
+  }
+
+  private static class CacheControlPolicy implements InstrumentationState {
+    private Optional<Integer> maxAge = Optional.empty();
+    private Optional<CacheControlScope> scope = Optional.of(CacheControlScope.PUBLIC);
+
+    void restrict(CacheControlPolicy policy) {
+      if (policy.maxAge.isPresent()
+          && (!maxAge.isPresent() || policy.maxAge.get() < maxAge.get())) {
+        this.maxAge = policy.maxAge;
+      }
+
+      if (policy.scope.isPresent()
+          && (!scope.isPresent() || !scope.get().equals(CacheControlScope.PRIVATE))) {
+        this.scope = policy.scope;
+      }
+    }
+
+    void restrict(Hint hint) {
```

Comment:
you could probably have Hint extend CacheControlPolicy in which case this duplicate method wouldn't be needed?